lib/galaxy/tools/actions/__init__.py

```diff
@@ -428,21 +429,19 @@ def handle_output(name, output, hidden=None):
 
                         effective_output_name = output_part_def.effective_output_name
                         element = handle_output(effective_output_name, output_part_def.output_def, hidden=True)
+                        created_element_datasets.append(element)
                         # TODO: this shouldn't exist in the top-level of the history at all
                         # but for now we are still working around that by hiding the contents
                         # there.
                         # Following hack causes dataset to no be added to history...
                         child_dataset_names.add(effective_output_name)
-
-                        history.add_dataset(element, set_hid=set_output_hid, quota=False)
                         trans.sa_session.add(element)
-                        trans.sa_session.flush()
-
                         current_element_identifiers.append({
                             "__object__": element,
                             "name": output_part_def.element_identifier,
                         })
 
+                    history.add_datasets(trans.sa_session, created_element_datasets, set_hid=True, quota=False, flush=True)
```

Comment:
Just to double check: you changed `set_hid` to be always `True` for the same reason explained in https://github.com/galaxyproject/galaxy/pull/6370/files#r197464679 , right?

Reply:
Good question - this time it wasn't intentional. I'll have to look into this more and see when set_hid would be false - hopefully it would never take this path in that case but I'll verify.

Reply:
> So I think this is only used for implicit converters which don't define explicit output collections so set_hid would never be false here.

Because we haven't implemented implicit converters for collections yet?

Changing it back to `set_output_hid` feels safer to me.

Reply:
> Because we haven't implemented implicit converters for collections yet?

No because implicit converters produce datasets - even when they work over collections - they are many individual jobs each producing a dataset. Converters can't produce collections - but I will change it back.